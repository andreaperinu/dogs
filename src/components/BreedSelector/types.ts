import * as O from 'fp-ts/lib/Option';

import { IBreed } from '@/models/models';

export interface IBreedSelectorProps {
	maybeSelectedBreed: O.Option<IBreed>,
	maybeBreeds: O.Option<IBreed[]>,
	onChange: (maybeBreed: O.Option<IBreed>) => void
}
