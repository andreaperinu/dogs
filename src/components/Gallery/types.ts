import * as O from 'fp-ts/lib/Option';

export interface IGalleryProps {
	maybeImages: O.Option<string[]>
}
