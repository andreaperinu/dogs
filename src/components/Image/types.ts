import * as O from 'fp-ts/lib/Option'

export interface IImageProps {
	maybeImage: O.Option<string>
	fixedSize?: boolean
}
