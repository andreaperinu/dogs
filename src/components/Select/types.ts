import * as O from 'fp-ts/lib/Option';
import { BaseComponent } from '../../types';

export interface ISelectProps extends BaseComponent {
	maybeValue: O.Option<ISelectItem>
	maybeOptions: O.Option<ISelectItem[]>
	maybeText?: O.Option<string>
	disabled?: boolean
	onChange: (item: O.Option<ISelectItem>) => void
}

export interface ISelectItem {
	id: string
	value: string
}

export interface ISelectStyledProps {
	showOptions: boolean
	disabled: boolean
}
