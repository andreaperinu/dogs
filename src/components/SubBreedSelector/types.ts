import * as O from 'fp-ts/lib/Option';

import { IBreed } from '@/models/models';

export interface ISubBreedSelectorProps {
	disabled?: boolean
	maybeSelectedSubBreed: O.Option<string>,
	maybeBreed: O.Option<IBreed>,
	onChange: (maybeSubBreed: O.Option<string>) => void
}

export interface ISubBreedSelectorStyledProps {
	disabled: boolean
}
