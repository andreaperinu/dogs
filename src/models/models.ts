export interface IBreed {
	breed: string
	subBreeds: string[]
}

export interface IResponse<T> {
	message?: T
	status: string
}

export type IBreedResponseMessage = {
	[breed: string]: string[]
}
