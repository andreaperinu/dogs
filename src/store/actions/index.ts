export { fetchBreeds, fetchRandomImage, resetRandomImage, fetchImages, resetImages } from './actions'
