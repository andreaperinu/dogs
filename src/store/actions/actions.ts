import {
	FETCH_BREEDS, FETCH_IMAGES, FETCH_RANDOM_IMAGE, RESET_IMAGES, RESET_RANDOM_IMAGE
} from '@/store/actions/rawActions';
import { Action } from '@/store/types';

export type BreedPayload = { breed: string }

export const fetchBreeds: Action = () => ({ type: FETCH_BREEDS })
export const fetchRandomImage: Action<BreedPayload> = (payload) => (
	{ type: FETCH_RANDOM_IMAGE, payload }
)

export const resetRandomImage: Action = () => ({ type: RESET_RANDOM_IMAGE })
export const resetImages: Action = () => ({ type: RESET_IMAGES })

export const fetchImages: Action<BreedPayload> = (payload) => ({ type: FETCH_IMAGES, payload })
