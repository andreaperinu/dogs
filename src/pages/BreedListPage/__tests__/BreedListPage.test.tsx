import * as A from 'fp-ts/lib/Array';
import { pipe } from 'fp-ts/lib/function';
import * as O from 'fp-ts/lib/Option';
import React from 'react';

import { IBreed } from '@/models/models';
import { BreedPayload } from '@/store/actions/actions';
import { FETCH_IMAGES } from '@/store/actions/rawActions';
import { IStoreState } from '@/store/types';
import * as mockedBreeds from '@/tests/mockedBreeds.json';
import { setup } from '@/tests/utils';
import { formatBreeds } from '@/utils/utils';

import component from '../BreedListPage';

const dispatch = jest.fn()

const maybeBreeds: O.Option<IBreed[]> = pipe(
	mockedBreeds,
	formatBreeds,
	O.fromNullable,
)

const initialState: Partial<IStoreState> = {
	maybeBreeds: O.none,
	maybeImages: O.none,
}

const maybeSelectedBreed: O.Option<IBreed> = pipe(maybeBreeds, O.chain(A.head))

describe('<BreedListPage />', () => {
	it('should match snapshot', () => {
		const wrapper = setup({ component, props: {} })
		expect(wrapper).toMatchSnapshot()
	})

	it('should dispatch fetch images', () => {
		const payload: BreedPayload = {
			breed: pipe(maybeSelectedBreed, O.map(({ breed }) => breed), O.getOrElse(() => ''))
		}

		jest.spyOn(React, 'useState').mockImplementation(() => [maybeSelectedBreed, jest.fn()])

		setup({ component, props: {}, initialState, dispatch, shouldMount: true, })

		expect(dispatch).toHaveBeenCalledWith({ payload, type: FETCH_IMAGES })
	})
})
