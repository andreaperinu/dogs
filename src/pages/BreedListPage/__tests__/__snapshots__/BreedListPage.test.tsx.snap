// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BreedListPage /> should match snapshot 1`] = `
<Fragment>
  <BreedSelector
    maybeSelectedBreed={
      Object {
        "_tag": "None",
      }
    }
    onChange={[Function]}
  />
  <Gallery />
</Fragment>
`;
